### PR TITLE
fix(userspace/engine): correctly handle evttype indexing corner cases 

### PR DESCRIPTION
**Pull request by jasondellaluce**
**What type of PR is this?**

/kind bug

**Any specific area of the project related to this PR?**

/area engine

**What this PR does / why we need it**:

Thanks to @incertum findings we discovered that the current event type resolver had some uncovered corner cases (see: https://github.com/falcosecurity/libs/pull/854#issuecomment-1411151732). This mostly involved cases where rules conditions have an `evt.type` combined with `not` negation, so it should not be a big deal, however fixing it is crucial for the correctness and predictability of rules' indexing.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
fix(userspace/engine): correctly handle evttype indexing corner cases 
```

### CODE REVIEW

File: tests/engine/test_filter_evttype_resolver.cpp [hunk 1]
```diff
@@ -214,7 +224,7 @@ TEST_CASE("Should find event types from filter", "[rule_loader]")
     SECTION("non_evt_type_or_nested_multi_not")
     {
         auto f = compile("evt.type=openat or not (evt.type=close and proc.name=nginx)");
-        compare_evttypes(std::move(f), not_close);
+        compare_evttypes(std::move(f), all_events);
```
Comment (jasondellaluce) [inline tests/engine/test_filter_evttype_resolver.cpp:227]: This test is where we were conceptually all wrong. In this case, the negated condition actually matches every event types, including `close`.

File: tests/engine/test_filter_evttype_resolver.cpp [hunk 2]
```diff
@@ -234,4 +244,127 @@ TEST_CASE("Should find event types from filter", "[rule_loader]")
...
+    SECTION("libs #854")
+    {
+        std::set<uint16_t> empty;
+        compare_filters_evttypes(
+            "evt.type in (connect, execve, accept, mmap, container) and not (proc.name=cat and evt.type=mmap)",
+            "evt.type in (accept, connect, container, execve, mmap)");
+        compare_evttypes(compile("(evt.type=mmap and not evt.type=mmap)"), empty);
+    }
+
+    SECTION("boolean algebra")
```
Comment (jasondellaluce) [inline tests/engine/test_filter_evttype_resolver.cpp:264]: Adding this test section allows us to assert that the set of event types correctly follows the boolean algebra semantics.

**Review by incertum — COMMENTED** [review]
:exploding_head: OMG Jason truly love it :heart: !
Simply extraordinary work and due diligence! Big applause :clap:  from my side!

Thanks so much for addressing these corner cases. Complexities involved here with negations and double-negations can be mind-bending for sure (also when writing Falco rules). Furthermore, the refactor improved the technical clarity by orders of magnitude!

**Review by FedeDP — APPROVED** [review]
/approve

**Review by Andreagit97 — APPROVED** [review]
/approve
